src/modules/ims_registrar_pcscf/service_routes.c

```diff
@@ -117,43 +120,81 @@ int checkcontact(struct sip_msg *_m, pcontact_t *c)
 {
 	int security_server_port = -1;
 	str received_host = {0, 0};
+	unsigned short received_port = 0;
+	char received_proto = 0;
 	char srcip[50];
 
+	if(trust_bottom_via) {
+		struct via_body *vb = cscf_get_last_via(_m);
+		if(vb == 0) {
+			LM_ERR("No via header in request\n");
+			return -1;
+		}
+		if(vb->received != NULL && vb->received->value.len > 0) {
+			received_host = vb->received->value;
+		} else {
+			received_host = vb->host;
+		}
+		if(vb->rport != NULL && vb->rport->value.len > 0) {
+			str2ushort(&vb->rport->value, &received_port);
+		} else {
+			received_port = vb->port;
+		}
+		// this probably doesn't matter, since IMS UEs register IPs for both
+		received_proto = vb->proto;
+	} else {
+		received_host.len = ip_addr2sbuf(&_m->rcv.src_ip, srcip, sizeof(srcip));
+		received_host.s = srcip;
+		received_port = _m->rcv.src_port;
+		received_proto = _m->rcv.proto;
+	}
+
 	LM_DBG("Port %d (search %d), Proto %d (search %d), reg_state %s (search "
 		   "%s)\n",
-			c->received_port, _m->rcv.src_port, c->received_proto,
-			_m->rcv.proto, reg_state_to_string(c->reg_state),
+			c->received_port, received_port, c->received_proto, received_proto,
+			reg_state_to_string(c->reg_state),
 			reg_state_to_string(PCONTACT_REGISTERED));
 
-	if(c->security) {
-		switch(c->security->type) {
-			case SECURITY_IPSEC:
-				security_server_port = c->security->data.ipsec->port_uc;
-				break;
-			case SECURITY_TLS:
-			case SECURITY_NONE:
-				break;
+
+	if(ipsec_pcscf.ipsec_on_expire == NULL) {
+		LM_DBG("ims_ipsec_pcscf module not loaded - skipping port-uc checks\n");
+	} else {
+		if(c->security) {
+			switch(c->security->type) {
+				case SECURITY_IPSEC:
+					security_server_port = c->security->data.ipsec->port_uc;
+					break;
+				case SECURITY_TLS:
+				case SECURITY_NONE:
+					break;
+			}
+		} else if(c->security_temp) {
+			switch(c->security_temp->type) {
+				case SECURITY_IPSEC:
+					security_server_port =
+							c->security_temp->data.ipsec->port_uc;
+					break;
+				case SECURITY_TLS:
+				case SECURITY_NONE:
+					break;
+			}
 		}
-	} else if(c->security_temp) {
-		switch(c->security_temp->type) {
-			case SECURITY_IPSEC:
-				security_server_port = c->security_temp->data.ipsec->port_uc;
-				break;
-			case SECURITY_TLS:
-			case SECURITY_NONE:
-				break;
+
+		if(!ignore_contact_rxport_check && (c->received_port == received_port)
+				&& (security_server_port == received_port)) {
+			LM_DBG("check contact failed - port-uc %d is neither contact "
+				   "received_port %d, nor message received port %d\n",
+					security_server_port, c->received_port, _m->rcv.src_port);
+			return 1;
 		}
 	}
 
 	if((ignore_reg_state || (c->reg_state == PCONTACT_REGISTERED))
+			// Weird... this condition is for rxport, not rxproto. If it was intentional, a comment would be nice,
+			// otherwise I'm thinking it's an unintended effect.
 			&& (ignore_contact_rxport_check
-					|| (c->received_port == _m->rcv.src_port)
```

Comment:
I am not sure about this change.. why was the check for SIP message received source port removed?

Reply:
It's moved to lines 183->184 in the additions.

I wish I hadn't, but since in my scenario the `port-uc` wasn't stored anywhere in Kamailio (`ims_ipsec_pcscf` not loaded), I had to put it behind an if for that module loaded. I'll make further improvement and also skip on `ignore_contact_rxport_check`. But it's still checked, just above.

Reply:
With `ignore_contact_rxport_check=1`  the protocol (not port! :upside_down_face:) hit me hard today, when REGISTER happened over TCP and then an MO MESSAGE kept being rejected because ... UDP.

I added a parameter `ignore_contact_rxproto_check` with default `1` (so changing behavior!). My opinion is that in IMS the IPsec SA is negotiated for all transport protocols (so in practice UDP and TCP), hence if a UE managed to correctly encrypt whatever UDP/TCP packet correctly and send it to us on the correct Security-Association flows, we should allow it.

Reply:
> My opinion is that in IMS the IPsec SA is negotiated for all transport protocols (so in practice UDP and TCP), hence if a UE managed to correctly encrypt whatever UDP/TCP packet correctly and send it to us on the correct Security-Association flows, we should allow it.

I agree